src/cast_core/src/CrudActor.cpp

```diff
@@ -1206,21 +1294,129 @@ struct CrudActor::PhaseConfig {
                                    id);
         };
 
-        operations = phaseContext.getPlural<std::unique_ptr<BaseOperation>>(
-            "Operation", "Operations", addOperation);
+        auto addState =
+            [&](const Node& node,
+                const std::unordered_map<std::string, int> states) -> std::unique_ptr<State> {
+            //           State myState;
+            auto stateName = node["Name"].to<std::string>();
+            // Skipping repeat for now
+            // operations
+            BOOST_LOG_TRIVIAL(debug) << "Adding state " << stateName;
+            auto stateOperations = node.getPlural<std::unique_ptr<BaseOperation>>(
+                "Operation", "Operations", addOperation);
+            BOOST_LOG_TRIVIAL(debug)
+                << "Got operations for state. Number of operations: " << stateOperations.size();
+            // Transitions
+            BOOST_LOG_TRIVIAL(debug) << "Got operations";
+            std::vector<double> weights;
+            std::vector<Transition> transitions;
+            for (auto [k, transitionYaml] : node["Transitions"]) {
+                weights.emplace_back(transitionYaml["Weight"].to<double>());
+                BOOST_LOG_TRIVIAL(debug)
+                    << "Next state name is " << transitionYaml["To"].to<std::string>();
+                transitions.emplace_back(Transition{
+                    states.at(transitionYaml["To"].to<std::string>()),
+                    Delay(transitionYaml["SleepBefore"], GeneratorArgs{phaseContext.rng(id), id})});
+            }
+            return std::unique_ptr<State>(new State{
+                std::move(stateOperations), stateName, std::move(weights), std::move(transitions)});
+        };
+        // Check if we have Operations or States. Through an error if we have both.
+        if (phaseContext["Operations"] && phaseContext["States"]) {
+            BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                "CrudActor has Operations and States at the same time."));
+        }
+        if (phaseContext["Operations"] || phaseContext["Operation"]) {
+            operations = phaseContext.getPlural<std::unique_ptr<BaseOperation>>(
+                "Operation", "Operations", addOperation);
+        } else if (phaseContext["States"]) {  // Parse out the states}
+            if (!phaseContext["InitialStates"]) {
+                BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                    "CrudActor has States, but has not specified InitialStates"));
+            }
+            BOOST_LOG_TRIVIAL(debug) << "Adding states";
+            // build the list of states, and then actuall process the states
+            int i = 0;
+            std::unordered_map<std::string, int> stateNames;
+            for (auto [k, state] : phaseContext["States"]) {
```

Comment:
consider a `States` class that encapsulates the name/id lookup and holds the `addState` results.

Reply:
If I pull the addStates out into a proper function, I also have to pull the addOperations out (since addStates calls addOperations). I started to do that and realize I have another problem. addOperation is called in this line of code

```
            auto stateOperations = node.getPlural<std::unique_ptr<BaseOperation>>(
                "Operation", "Operations", addOperation);

```

I don't know how to pass the added arguments through that. The lambda function is allowing us to pass things through the environment. I can't do that with a proper function. If @rtimmons has an answer for that problem I'll put out addOperations and addStates into separate functions in the anonymous namespace above, replacing the two lambda functions.